# Parameters


* `index_cpus` (number of cpus reserved for the genome indexation process.   `default=14`)
* `mapping_cpus` (idem. for the mapping process, used to create BAM files. `default=14`)
* `counting_cpus` (idem. for the counting process. `default=7`)
* `mapping_memory` (RAM reserved for mapping. `default=50GB`)


If you already possess some of the files needed to execute the pipeline, you can specify them as follows:

* `reads` (path pointing to a directory containing the `fasterq` files)
* `genome` (path pointing to a directory containing the genome FASTA file)
* `index` (Répertoire contenant les fichiers d’index)
* `mapping` (Répertoire contenant les fichiers BAM)
* `counting` (Chemin d’accès entier au fichier de comptage – comprend le fichier lui-même)
* `metadata` (Chemin d’accès entier au fichier de métadonnées – comprend le fichier lui-même)

If unspecified, the pipeline will be executed using default values from the config file : [nextflow.config](https://github.com/bio-TAGI/Hackathon/blob/main/Nextflow/nextflow.config). These too, can be tweaked and overriden:

* `ids` List of SRR accession number to fetch paired-end fastq files. 
  * default=`['SRR628582', 'SRR628583', 'SRR628584', 'SRR628585', 'SRR628586', 'SRR628587', 'SRR628588', 'SRR628589']`
* `genome_url` URL to download the reference genome. 
  *  default `ftp://ftp.ensembl.org/pub/release-101/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz`
* `annotation_url` URL to donwload the reference genome's annotation. 
  * default `ftp://ftp.ensembl.org/pub/release-101/gtf/homo_sapiens/Homo_sapiens.GRCh38.101.chr.gtf.gz`
* `sjdbOverhang` (a STAR-specific parameter. `default=99`)
  * For further information about this parameter, see [this tutorial](https://sydney-informatics-hub.github.io/training-RNAseq/02-BuildAGenomeIndex/index.html), 
  or the [STAR manual](https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf).
